### PR TITLE
No unguarded window

### PR DESCRIPTION
**Pull request by scalvert**
@chadhietala 

## Description

Adding no-unguarded-window rule + tests

## Changes

- Added `no-unguarded-window` rule
- Added tests

### CODE REVIEW

File: guides/rules/no-unguarded-window.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+To protect against errors during server side rendering, you should ensure you're guarding the usages
+of `window` with a check to `environment.isBrowser()`.
+
+Example:
+
+```js
+import Ember from 'ember';
+
+Ember.component.extend({
```
Comment (knownasilya) [inline guides/rules/no-unguarded-window.md]: `Ember.Component`

File: guides/rules/no-unguarded-window.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  init() {
+    if (environment.isBrowser()) {
```
Comment (knownasilya) [inline guides/rules/no-unguarded-window.md]: Where is `environment` coming from?

**Comment by stefanpenner** [comment]
> Added no-unguarded-window rule

is this correct? There are some hooks that window is fine to use in, such as the DOM only hooks in components. Is the hinting aware of this?

Should we consider Banning window altogether and telling people to use `self` instead? As in web workers/service workers there is only `self` and no `window`...

**Comment by scalvert** [comment]
@stefanpenner that's a fair point. In the `no-timers` rule we fire it if it's not in actions. What hooks specifically would it be safe to use `window`?

**Comment by stefanpenner** [comment]
> What hooks specifically would it be safe to use window?

i don't recall, but didInsertElement/willDestroyElement and likely actions (but that isn't firm)

along these lines, `this.$()` and `this.element`   are only useful in some context (only if the element is present.)

**Comment by scalvert** [comment]
Yap, makes sense. I'm going to check the Ember source to see what hooks are called in FastBoot, and will update/add checks accordingly.

File: package.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
     "test": "tests"
   },
   "scripts": {
-    "test": "mocha tests tests/lib/rules/*.js tests/**/*-test.js"
+    "test": "mocha --recursive tests"
```
Comment (chadhietala) [inline package.json]: Neat

File: lib/rules/linkedin/no-unguarded-window.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+function isWindowReference(node) {
+  return node.object.name === 'window';
+}
+
+module.exports = {
+  meta: {
+    message: MESSAGE
+  },
+  create(context) {
+    return {
```
Comment (chadhietala) [inline lib/rules/linkedin/no-unguarded-window.js]: Can you capture the import bindings and do the check. We ideally don't want to create reserved worlds or members. Failing case would be:

```
import Ember from 'ember';
let trollSteve = {
  isBrowser() { return 'tacos' }
}

export default Ember.Component.extend({
  init() {
   this._super(...arguments);
   if (trollSteve.isBrowser()) {
    let allTheTacos = true;
   }
  }
})
```
Also this is a case we want to cover and is legal.
```
import Ember from 'ember';
import environment from 'ember-stdlib/utils/environment';
const { isBrowser } = environment 

export default Ember.Component.extend({
  init() {
   this._super(...arguments);
   if (isBrowser()) {
    let allTheTacos = true;
   }
  }
})
```

File: package.json (hunk 3 above)
Comment (scalvert) [inline package.json]: :)

**Pull request merged**